firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -87,6 +108,259 @@ public void addFieldIndex(FieldIndex index) {
+      }
+      values.add(field);
+    }
+    return values;
+  }
+
+  @Override
+  @Nullable
+  public Set<DocumentKey> getDocumentsMatchingTarget(Target target) {
+    @Nullable FieldIndex fieldIndex = getMatchingIndex(target);
+    if (fieldIndex == null) return null;
+
+    Bound lowerBound = target.getLowerBound(fieldIndex);
+    @Nullable Bound upperBound = target.getUpperBound(fieldIndex);
+
+    if (Logger.isDebugEnabled()) {
+      Logger.warn(
+          TAG,
+          "Using index '%s' to execute '%s' (Lower bound: %s, Upper bound: %s)",
+          fieldIndex,
+          target,
+          lowerBound,
+          upperBound);
+    }
+
+    Set<DocumentKey> result = new HashSet<>();
+
+    if (upperBound != null) {
+      List<byte[]> lowerBoundValues =
+          encodeTargetValues(fieldIndex, target, lowerBound.getPosition());
+      List<byte[]> upperBoundValues =
+          encodeTargetValues(fieldIndex, target, upperBound.getPosition());
+
+      hardAssert(
+          lowerBoundValues.size() == upperBoundValues.size(),
+          "Expected upper and lower bound size to match");
+
+      // TODO(indexing): To avoid reading the same documents multiple times, we should ideally only
+      // send one query that combines all clauses.
+      for (int i = 0; i < lowerBoundValues.size(); ++i) {
+        db.query(
+                String.format(
+                    "SELECT document_name from index_entries WHERE index_id = ? AND index_value %s ? AND index_value %s ?",
+                    lowerBound.isBefore() ? ">=" : ">", upperBound.isBefore() ? "<=" : "<"))
+            .binding(fieldIndex.getIndexId(), lowerBoundValues.get(i), upperBoundValues.get(i))
+            .forEach(
+                row -> result.add(DocumentKey.fromPath(ResourcePath.fromString(row.getString(0)))));
+      }
+    } else {
+      List<byte[]> lowerBoundValues =
+          encodeTargetValues(fieldIndex, target, lowerBound.getPosition());
+      for (byte[] lowerBoundValue : lowerBoundValues) {
+        db.query(
+                String.format(
+                    "SELECT document_name from index_entries WHERE index_id = ? AND index_value %s  ?",
+                    lowerBound.isBefore() ? ">=" : ">"))
+            .binding(fieldIndex.getIndexId(), lowerBoundValue)
+            .forEach(
+                row -> result.add(DocumentKey.fromPath(ResourcePath.fromString(row.getString(0)))));
+      }
+    }
+
+    Logger.debug(TAG, "Index scan returned %s documents", result.size());
+    return result;
+  }
+
+  /**
+   * Returns an index that can be used to serve the provided target. Returns {@code null} if no
+   * index is configured.
+   */
+  private @Nullable FieldIndex getMatchingIndex(Target target) {
+    TargetIndexMatcher targetIndexMatcher = new TargetIndexMatcher(target);
+    String collectionGroup =
+        target.getCollectionGroup() != null
+            ? target.getCollectionGroup()
+            : target.getPath().getLastSegment();
```

Comment:
I am confused here, if we use `getLastSegment` here, would it cause conflicts? I was under the impression that `collection_group` is full path?

Reply:
No, the collectionGroup is the name of the parent collection.